File: modules/veloq-widget/android/src/main/java/com/veloq/widget/VeloqWidgetModule.kt
package com.veloq.widget

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

// Infrastructure constant (not theme). The AppWidgetProvider reads the same file from
// the app's private filesDir — provider and module run in the same app process.
private const val SNAPSHOT_FILE = "widget-snapshot.json"

/**
 * Bridges the JS snapshot pipeline to the Android home-screen widget: writes the
 * pre-formatted JSON to the app's filesDir (where the provider reads it), then
 * broadcasts an update to every widget provider this app declares so each instance
 * redraws. The provider never computes — it only renders this file.
 */
class VeloqWidgetModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("VeloqWidget")

    Function("writeSnapshot") { json: String ->
      // Write to a temp file then rename so a concurrent onUpdate never reads a
      // half-written snapshot (mirrors the iOS `.atomic` write). renameTo within
      // filesDir is atomic on the app's internal storage; fall back to a direct
      // write only if the rename is refused.
      val dir = context.filesDir
      val tmp = File(dir, "$SNAPSHOT_FILE.tmp")
      val dest = File(dir, SNAPSHOT_FILE)
      tmp.writeText(json)
      if (!tmp.renameTo(dest)) {
        dest.writeText(json)
        tmp.delete()
      }
    }

    Function("reloadWidgets") {
      val ctx = context.applicationContext
      val manager = AppWidgetManager.getInstance(ctx)
      val pkg = ctx.packageName
      // Notify whatever widget providers this app installs, without hardcoding the
      // provider class (its FQN diverges from applicationId under the dev/prod split).
      for (info in manager.installedProviders) {
        if (info.provider.packageName != pkg) continue
        val ids = manager.getAppWidgetIds(info.provider)
        if (ids.isEmpty()) continue
        val intent =
          Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE).apply {
            component = info.provider
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
          }
        ctx.sendBroadcast(intent)
      }
    }
  }
}

File: modules/veloq-widget/ios/VeloqWidgetModule.swift
import ExpoModulesCore
import WidgetKit

// Shared-container coordinates. These are infrastructure constants (not theme), so
// they live here and are mirrored by the widget extension's snapshot reader. The App
// Group is fixed across dev/prod, matching the iCloud container in with-icloud.js.
private let kAppGroup = "group.com.veloq.app"
private let kSnapshotFile = "widget-snapshot.json"

/// Bridges the JS snapshot pipeline to WidgetKit: writes the pre-formatted JSON into
/// the App Group container the widget extension reads, then asks WidgetKit to redraw.
/// The widget itself never computes — it only renders this file.
public final class VeloqWidgetModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VeloqWidget")

    Function("writeSnapshot") { (json: String) in
      guard
        let dir = FileManager.default.containerURL(
          forSecurityApplicationGroupIdentifier: kAppGroup
        )
      else { return }
      let url = dir.appendingPathComponent(kSnapshotFile)
      try? Data(json.utf8).write(to: url, options: .atomic)
    }

    Function("reloadWidgets") {
      WidgetCenter.shared.reloadAllTimelines()
    }
  }
}
